cli/maintainer/src/maintenance.rs

```diff
@@ -754,34 +773,87 @@ impl SolidoState {
+        let anker_state = self.anker_state.as_ref()?;
 
+        let rewards = self.get_anker_rewards()?;
         let min_rewards_to_sell = self
             .solido
             .exchange_rate
             .exchange_sol(Self::MINIMUM_WITHDRAW_AMOUNT)
             .expect("The price of a signature should be small enough that it doesn't overflow.");
+        // We should not call the instruction if the rewards are 0, or if the rewards are so small
+        // that the transaction cost is a significant portion of the rewards.
+        if rewards < min_rewards_to_sell {
+            return None;
+        }
+
+        let expected_proceeds = anker_state
+            .constant_product_calculator
+            .swap_without_fees(
+                rewards.0 as u128,
+                anker_state.pool_st_sol_balance.0 as u128,
+                anker_state.pool_ust_balance.0 as u128,
+                TradeDirection::AtoB,
+            )?
+            .destination_amount_swapped;
+        let expected_proceeds = MicroUst(expected_proceeds as u64);
 
-        // Estimate the amount of UST that we will get when swapping against the constant product
-        // pool. This is only an approximation because it doesn’t uphold the constant product
-        // invariant, but this is a reasonable approximation when the rewards to sell are small
-        // compared to the pool balance. Also, this ignores swap fees. It would have been nice
-        // to use the actual `spl_token_swap::state::SwapV1` instance, but it is not `Send`, so
-        // we can’t store it in our state.
-        let ust_per_st_sol = Rational {
-            numerator: anker_state.pool_ust_balance.0,
-            denominator: anker_state.pool_st_sol_balance.0,
-        };
-        let expected_proceeds = (rewards * ust_per_st_sol)
-            .map(|x| MicroUst(x.0))
-            .unwrap_or(MicroUst(0));
         // We want at least 0.01 UST out if we are going to do the swap at all.
         let min_proceeds = MicroUst(10_000);
 
-        // We should not call the instruction if the rewards are 0, or if the rewards are so small
-        // that the transaction cost is a significant portion of the rewards.
-        if rewards < min_rewards_to_sell || expected_proceeds < min_proceeds {
-            None
+        // We want at least 0.01 UST out if we are going to do the swap at all.
+        if expected_proceeds < min_proceeds {
+            return None;
+        }
+
+        // Check if we can sell the rewards with the preset slippage tolerance.
+        // Note that this might change when the instruction gets included in the block.
+        let minimum_ust_amount_for_rewards = anker_state
+            .anker
+            .historical_st_sol_prices
+            .minimum_ust_swap_amount(rewards, anker_state.anker.sell_rewards_min_out_bps)
+            .ok()?;
+        if expected_proceeds < minimum_ust_amount_for_rewards {
```

Comment:
I think this check is a bit misleading, because `expected_proceeds` is computed through `swap_without_fees`, but the `minimum_ust_swap_amount` is including the fees. So the `expected_proceeds` is an overestimation.

How difficult would it be to take the swap fees into account when computing `expected_proceeds`? If I recall correctly it’s just a fixed percentage, so if it’s one additional multiplication by a `Rational`, that should be doable.

Reply:
But at `get_one_st_sol_for_ust_price_from_pool` https://github.com/ChorusOne/solido/blob/7009abc6ba37063aaeb0a2ef3b374f4f20cb17b3/anker/src/logic.rs#L316 it's also swapping without fees, am I missing something?

Reply:
Yes, it is, but the `sell_rewards_min_out_bps` is the tolerance for slippage + fees. So if we set the `sell_rewards_min_out_bps` to 9800 bps (98%), then if the fees are 0.1% we would tolerate 1.9% slippage, but if the fees are 1%, we would tolerate only 1% slippage.

Reply:
Apparently the swap takes fees from the `source`:
https://github.com/solana-labs/solana-program-library/blob/95cd64ce022844d5ad2b2d4809c4b29c063162b3/token-swap/program/src/curve/base.rs#L80-L84
The swapped amount is going to remain the same, just some of the rewards are going to be `fees`.

Reply:
They compute the fees from the source amount, then subtract that from the source amount, and then put that in `swap_without_fees`. So the end result is still less output, then when you would put the full input amount in `swap_without_fees`.

Reply:
Indeed, I added the `Fees` struct to the `AnkerState` and subtracted the trade and owner fee from the rewards. The alternative is to either save the swap pool or store a closure in the struct.
The swap pool cannot be saved trivially in the struct because it has a trait object that doesn't implement `Send + Sync` that's needed in another part of the code.
I started adding the closure, but the changes propagate all throughout the code if I added it as a generic argument. So I think this is the simplest way of doing it, it's just a couple of lines.